File: python/compile_lens/_schema.py
"""Pydantic v2 models for the compile-lens ``.cls.json`` session artifact (schema v0.5.0).

This is the **Python half** of the cross-language contract; the Rust half is the
``cls-schema`` crate. The two must produce/consume identical JSON so the cross-language
round-trip test (Python writes -> Rust reads -> consistent, and reverse) passes.
These models mirror ``schema/v0.5.0.json`` and ``crates/cls-schema/src/*.rs`` 1:1; the
per-decision rationale lives in the comments below.

Representation conventions (mirroring the Rust side):

- **Numbers**: schema ``integer`` -> ``int``, ``number`` -> ``float`` (incl. ``flops`` /
  ``bytes_*``, which arrive in scientific notation), ``string`` -> ``str``.
- **Unknown-field capture (DR4)**: only :class:`ClsArtifact` and :class:`Session` use
  ``extra="allow"`` (unknown keys preserved through a round trip). Every other model uses
  ``extra="ignore"`` (unknown keys accepted on read, dropped on write).
- **Optional vs. nullable (DR3)**: serialization uses ``exclude_unset=True``
  (see :func:`to_json`), so a key absent from the input stays absent and a present ``null``
  stays ``null``. The two schema ``[T, null]`` union fields
  (:attr:`CompileConfig.dynamic`, :attr:`EnvSnapshot.random_seed`) are meaningful when
  ``null``; a producer constructing a model from scratch must pass them explicitly so they
  are emitted (see their field docstrings).
- **Fail-closed enum (DR6)**: :class:`RedactionPolicy` is the only constrained string
  modeled as an enum; an unknown value raises ``ValidationError``. Other constrained
  strings stay ``str`` for forward compatibility (D6).
- **Determinism (DR5)**: Python ``dict`` preserves insertion order, so open maps need no
  special handling (the Rust side uses ``IndexMap`` for the same reason). Never serialize
  with ``sort_keys``.
"""

from __future__ import annotations

import math
from enum import StrEnum
from typing import Annotated, Any

from pydantic import BaseModel, BeforeValidator, ConfigDict, Field, PlainSerializer

#: Schema version these models mirror.
SCHEMA_VERSION: str = "0.5.0"


def _decode_non_finite(value: Any) -> Any:
    """Read side: map the JSON string sentinels back to floats; pass numbers through. (pydantic also
    coerces these by default; doing it explicitly keeps the contract symmetric with the writer and
    with the Rust adapter, and lets any *other* string be rejected rather than silently coerced.)"""
    if isinstance(value, str):
        return {"NaN": math.nan, "Infinity": math.inf, "-Infinity": -math.inf}.get(value, value)
    return value


def _encode_non_finite(value: float) -> float | str:
    """Write side: a non-finite float becomes an explicit string sentinel; a finite one stays a
    JSON number."""
    if math.isnan(value):
        return "NaN"
    if math.isinf(value):
        return "Infinity" if value > 0 else "-Infinity"
    return value


#: A ``float`` that encodes non-finite values (NaN, ±∞) as the explicit JSON string sentinels
#: ``"NaN"`` / ``"Infinity"`` / ``"-Infinity"`` instead of the silent ``null`` both serializers
#: default to — so a NaN survives the Python→Rust round trip and stays distinct from "not computed".
#: Applied to every nullable float that holds a computed or measured value. See ADR-039.
NonFiniteFloat = Annotated[
    float,
    BeforeValidator(_decode_non_finite),
    PlainSerializer(_encode_non_finite, return_type=float | str, when_used="json"),
]


class _Model(BaseModel):
    """Shared base: unknown keys are accepted but not preserved (``extra="ignore"``).

    The two growth-point models override this with ``extra="allow"``.
    """

    model_config = ConfigDict(extra="ignore")


class RedactionPolicy(StrEnum):
    """Redaction classification (D11). Closed + fail-closed: an unknown wire value raises
    ``ValidationError`` rather than silently defaulting to a more permissive policy."""

    DEFAULT_STRICT = "default-strict"
    INTERNAL = "internal"
    CONFIDENTIAL = "confidential"
    PUBLIC_SAFE = "public-safe"


# ── shared location types ───────────────────────────────────────────────────────────────
class SourceLocation(_Model):
    file: str | None = None
    line: int | None = None
    function: str | None = None


class SourceRange(_Model):
    file: str | None = None
    line_start: int | None = None
    line_end: int | None = None
    code_excerpt: str | None = None


# ── records (mirror records.rs) ─────────────────────────────────────────────────────────
class FailedGuard(_Model):
    guard_id: str | None = None
    expression: str | None = None
    previous_value: str | None = None
    new_value: str | None = None


class GuardEvaluation(_Model):
    guard_id: str | None = None
    evaluation_result: str | None = None
    evaluated_value: str | None = None


class InternalStateSnapshot(_Model):
    module_buffers_hash: str | None = None
    module_attrs_changed: list[str] = Field(default_factory=list)


class ReferenceIssue(_Model):
    url: str | None = None
    fixed_in_version: str | None = None


class GraphBreak(_Model):
    break_id: str
    reason: str
    location: SourceLocation | None = None
    op_or_construct: str | None = None
    fallback_kind: str | None = None


class Recompilation(_Model):
    recompilation_id: str
    compiled_function_id: str | None = None
    trigger_reason: str | None = None
    failed_guard: FailedGuard | None = None
    occurred_at_step: int | None = None
    wall_clock_ms: NonFiniteFloat | None = None


class FxNode(_Model):
    #: Graph-unique node id; other nodes reference it in their ``inputs`` to encode an edge.
    id: str
    #: The operation, e.g. ``aten.matmul`` or a ``call_function`` target.
    op_type: str
    #: Ordered ids of the upstream nodes this one consumes. Order is load-bearing:
    #: ``sub(a, b)`` and ``sub(b, a)`` differ only here, so operand-order regressions are
    #: detectable only because the sequence is preserved rather than treated as a set.
    inputs: list[str] = Field(default_factory=list)
    #: Node attributes (a constant's value, a dim parameter, …); a difference here makes the
    #: diff classify an otherwise-identical node as ``modified``.
    attrs: dict[str, Any] = Field(default_factory=dict)
    #: Output tensor shape, when a concrete one was captured from ``node.meta['val']``. Empty when
    #: unknown (meta absent, value not a single tensor, or a dynamic/symbolic shape). Tool 6's cost
    #: model reads these to size HBM traffic; structure-only consumers ignore them (ADR-038).
    out_shape: list[int] = Field(default_factory=list)
    #: Output dtype as a torch name without the ``torch.`` prefix (``bfloat16``); absent under the
    #: same conditions as ``out_shape``.
    out_dtype: str | None = None


class CompiledGraph(_Model):
    graph_id: str
    compiled_function_id: str | None = None
    fx_graph_path: str | None = None
    inductor_ir_path: str | None = None
    guard_list: list[str] = Field(default_factory=list)
    kernel_ids: list[str] = Field(default_factory=list)
    compile_phases_summary: dict[str, Any] = Field(default_factory=dict)
    #: Node-level FX graph structure for the WL-signature diff (Tool 2a, ADR-024). Inlined so
    #: a single ``.cls.json`` stays self-contained; optional and forward-compatible.
    nodes: list[FxNode] = Field(default_factory=list)


class CompilePhase(_Model):
    name: str
    duration_ms: NonFiniteFloat | None = None


class Iteration(_Model):
    iteration_index: int
    timestamp_ms: NonFiniteFloat | None = None
    guard_evaluations: list[GuardEvaluation] = Field(default_factory=list)
    cache_hit: bool | None = None
    recompilation_triggered: bool | None = None
    output_signature: str | None = None
    internal_state_snapshot: InternalStateSnapshot | None = None


class LintFinding(_Model):
    finding_id: str
    pattern_category: str  # plain str (vocabulary grows post-v1.0); not an enum (D6)
    severity: str
    source_location: SourceRange | None = None
    trigger_pattern: str | None = None
    li_et_al_section: str | None = None
    reference_issue: ReferenceIssue | None = None
    workaround: str | None = None
    confidence: str | None = None
    applies_to_user_torch_version: bool | None = None


# ── divergence (Tool 3; mirror records.rs Divergence / DivergenceAttribution, ADR-034) ───
class DivergenceAttribution(_Model):
    """Pass-level causal attribution for a divergence (ADR-032 / ADR-034). Pass-level only:
    ``responsible_passes`` names the inductor pass(es) whose disabling removes the divergence,
    never a fused node."""

    attributed: bool
    responsible_passes: list[str] = Field(default_factory=list)
    summary: str
    num_probes: int


class Divergence(_Model):
    """One eager-vs-compiled divergence finding (``divergences[]``). An *analysis result*, so a
    normalized array record like ``lint_findings`` (ADR-021/034); the attribution is nested.
    ``first_divergent_layer`` is absent when nothing diverged; ``max_abs_diff`` is absent for a
    shape mismatch or no divergence."""

    divergence_id: str
    first_divergent_layer: str | None = None
    max_abs_diff: NonFiniteFloat | None = None
    num_layers_compared: int
    rtol: float
    atol: float
    suggested_cause: str | None = None
    attribution: DivergenceAttribution | None = None


# ── kernels + roofline (mirror kernels.rs) ──────────────────────────────────────────────
class LaunchConfig(_Model):
    grid: list[int] = Field(default_factory=list)
    block: list[int] = Field(default_factory=list)
    shared_mem_bytes: int | None = None
    num_warps: int | None = None
    num_stages: int | None = None


class KernelFeatures(_Model):
    flops: NonFiniteFloat | None = None  # schema `number` -> float (arrives as 1.2e10 etc.)
    bytes_loaded: NonFiniteFloat | None = None
    bytes_stored: NonFiniteFloat | None = None
    block_size: int | None = None
    num_regs: int | None = None  # register pressure proxy (Tool 5 4th correction)
    n_spills: int | None = None


class KernelMeasurements(_Model):
    iterations: int | None = None
    mean_us: NonFiniteFloat | None = None
    median_us: NonFiniteFloat | None = None
    p99_us: NonFiniteFloat | None = None
    source: str | None = None  # plain str ("proton"/"self_timed"); forward-compat


class Kernel(_Model):
    kernel_id: str
    name: str
    source_path: str | None = None
    ptx_path: str | None = None
    kernel_source_excerpt: str | None = None  # opt-in only
    launch_config: LaunchConfig | None = None
    occupancy_hint: NonFiniteFloat | None = None
    features: KernelFeatures | None = None
    measurements: KernelMeasurements | None = None
    autotune_history: list[dict[str, Any]] = Field(default_factory=list)


class GpuSpec(_Model):
    name: str | None = None
    peak_compute_tflops_bf16: NonFiniteFloat | None = None
    peak_bw_gbps: NonFiniteFloat | None = None
    l2_cache_mb: NonFiniteFloat | None = None
    sm_count: int | None = None


class Calibration(_Model):
    pearson_correlation: NonFiniteFloat | None = None  # float (may be negative)
    spearman_correlation: NonFiniteFloat | None = None
    calibration_verdict: str | None = None
    n_samples: int | None = None


class PruningDecision(_Model):
    mode: str | None = None  # plain str; forward-compat
    threshold_rank_correlation: NonFiniteFloat | None = None
    configs_total: int | None = None
    configs_predicted_top_k: int | None = None
    configs_measured: int | None = None


class Suggestion(_Model):
    text: str | None = None
    evidence_link: str | None = None


class RooflinePrediction(_Model):
    kernel_id: str
    gpu_spec: GpuSpec | None = None
    arithmetic_intensity: NonFiniteFloat | None = None
    bound_type: str | None = None  # plain str ("memory"/"compute"); forward-compat
    theoretical_lower_bound_us: NonFiniteFloat | None = None
    empirical_predictor_us: NonFiniteFloat | None = None
    measured_us: NonFiniteFloat | None = None
    achieved_vs_predictor: NonFiniteFloat | None = None
    corrections_applied: list[str] = Field(default_factory=list)
    calibration: Calibration | None = None
    pruning_decision: PruningDecision | None = None
    verdict: str | None = None
    suggestions: list[Suggestion] = Field(default_factory=list)


# ── session (mirror session.rs) ─────────────────────────────────────────────────────────
class CompileConfig(_Model):
    fullgraph: bool | None = None
    #: ``torch.compile(dynamic=...)``. Schema type ``[boolean, null]``: ``null`` means
    #: "auto" and is a distinct, meaningful state. On a round trip (parsed input) a present
    #: ``null`` is preserved by ``exclude_unset``. A producer building this from scratch
    #: must pass ``dynamic=`` explicitly (even ``None``) for it to be emitted.
    dynamic: bool | None = None
    backend: str | None = None
    mode: str | None = None
    options: dict[str, Any] = Field(default_factory=dict)  # open map; cannot be flattened


class EnvSnapshot(_Model):
    relevant_env_vars: dict[str, str] = Field(default_factory=dict)
    #: Schema type ``[integer, null]``: a present-``null`` seed (deliberately unset) differs
    #: from "no seed field". Same construct-vs-parse caveat as :attr:`CompileConfig.dynamic`.
    random_seed: int | None = None


class Session(_Model):
    """Run-level metadata. Growth point: ``extra="allow"`` preserves unknown session keys
    (e.g. a future ``gpu_arch``) through a round trip, mirroring the Rust ``Session.extra``."""

    model_config = ConfigDict(extra="allow")

    id: str
    timestamp: str
    torch_version: str
    redaction_policy: RedactionPolicy
    triton_version: str | None = None
    cuda_version: str | None = None
    python_version: str | None = None
    gpu_name: str | None = None
    host: str | None = None
    command: str | None = None
    duration_ms: int | None = None
    iteration_count: int | None = None
    git_sha: str | None = None
    workspace_fingerprint: str | None = None
    rank: int | None = None
    world_size: int | None = None
    collector_versions: dict[str, str] = Field(default_factory=dict)
    compile_config: CompileConfig | None = None
    env_snapshot: EnvSnapshot | None = None


class FusionShape(_Model):
    m: int | None = None
    n: int | None = None
    k0: int | None = None
    k1: int | None = None
    dtype: str | None = None


class FusionLocation(_Model):
    fx_node_ids: list[str] = Field(default_factory=list)
    src_lineno_range: str | None = None


class FusionOpportunity(_Model):
    """Tool 6 algebraic fusion opportunity (``fusion_opportunities[]``). Suggest-only — the
    detector identifies and quantifies the opportunity and describes the fusion to apply in
    ``suggested_fusion``; it never writes, imports, or runs a kernel."""

    pattern_id: str
    location: FusionLocation | None = None
    shape: FusionShape | None = None
    baseline_hbm_bytes: NonFiniteFloat | None = None
    fused_hbm_bytes: NonFiniteFloat | None = None
    estimated_speedup: NonFiniteFloat | None = None
    suggested_fusion: str | None = None
    confidence: str | None = None  # "high" / "medium" / "low" (plain str, forward-compat)


class ClsArtifact(_Model):
    """Top-level ``.cls.json`` artifact. Growth point: ``extra="allow"`` preserves unknown
    top-level keys (e.g. a future record array), mirroring the Rust ``ClsArtifact.extra``."""

    model_config = ConfigDict(extra="allow")

    schema_version: str
    session: Session
    graph_breaks: list[GraphBreak] = Field(default_factory=list)
    recompilations: list[Recompilation] = Field(default_factory=list)
    compiled_graphs: list[CompiledGraph] = Field(default_factory=list)
    compile_phases: list[CompilePhase] = Field(default_factory=list)
    iterations: list[Iteration] = Field(default_factory=list)
    lint_findings: list[LintFinding] = Field(default_factory=list)
    divergences: list[Divergence] = Field(default_factory=list)
    kernels: list[Kernel] = Field(default_factory=list)
    roofline_predictions: list[RooflinePrediction] = Field(default_factory=list)
    fusion_opportunities: list[FusionOpportunity] = Field(default_factory=list)


def from_json(data: str | bytes) -> ClsArtifact:
    """Parse a ``.cls.json`` document into a :class:`ClsArtifact`."""
    return ClsArtifact.model_validate_json(data)


def to_json(artifact: ClsArtifact) -> str:
    """Serialize a :class:`ClsArtifact` back to JSON.

    Uses ``exclude_unset=True`` so the output mirrors serde's behavior: keys that were
    absent in the parsed input stay absent, while a present ``null`` (e.g. ``dynamic``)
    is preserved. This is the faithful round-trip path the Rust side reads back.
    """
    return artifact.model_dump_json(exclude_unset=True)

File: crates/cls-schema/tests/schema.rs
//! Integration tests for `cls-schema`.
//!
//! These exercise the crate through its public API exactly as a downstream analyzer would,
//! and they deserialize the *canonical* example artifacts shipped in `schema/examples/` so
//! the Rust bindings and the JSON Schema can never silently drift apart.

use cls_schema::{ClsArtifact, RedactionPolicy, Session, SCHEMA_VERSION};

/// Canonical minimal artifact (only the required fields). Source of truth = JSON Schema
/// examples; included at compile time so a schema edit that breaks the bindings fails the
/// build.
const MINIMAL: &str = include_str!("../../../schema/examples/minimal.cls.json");
/// Canonical fully-populated artifact (every field exercised at least once).
const FULL: &str = include_str!("../../../schema/examples/full.cls.json");

fn parse(json: &str) -> ClsArtifact {
    serde_json::from_str(json).expect("example artifact should deserialize")
}

// ── test_minimal_example_deserialize ────────────────────────────────────────────────────
#[test]
fn test_minimal_example_deserialize() {
    let artifact = parse(MINIMAL);

    assert_eq!(artifact.schema_version, SCHEMA_VERSION);
    assert_eq!(artifact.session.id, "00000000-0000-4000-8000-000000000000");
    assert_eq!(artifact.session.torch_version, "2.6.0");
    assert_eq!(
        artifact.session.redaction_policy,
        RedactionPolicy::DefaultStrict
    );

    // Everything optional is absent, and the record arrays default to empty (not null).
    assert!(artifact.session.compile_config.is_none());
    assert!(artifact.session.env_snapshot.is_none());
    assert!(artifact.session.triton_version.is_none());
    assert!(artifact.graph_breaks.is_empty());
    assert!(artifact.kernels.is_empty());
    assert!(artifact.session.extra.is_empty());
}

// ── test_full_example_deserialize ───────────────────────────────────────────────────────
#[test]
fn test_full_example_deserialize() {
    let artifact = parse(FULL);

    let session = &artifact.session;
    assert_eq!(session.gpu_name.as_deref(), Some("NVIDIA H100 80GB HBM3"));
    assert_eq!(session.duration_ms, Some(45_230));
    assert_eq!(session.world_size, Some(1));
    assert_eq!(
        session
            .collector_versions
            .get("recompile")
            .map(String::as_str),
        Some("0.5.0")
    );

    // Nested [T, null] union field: present and explicitly null.
    let compile_config = session
        .compile_config
        .as_ref()
        .expect("compile_config present");
    assert_eq!(compile_config.fullgraph, Some(false));
    assert_eq!(compile_config.dynamic, None);
    assert_eq!(compile_config.backend.as_deref(), Some("inductor"));

    let env = session.env_snapshot.as_ref().expect("env_snapshot present");
    assert_eq!(env.random_seed, Some(42));
    assert_eq!(
        env.relevant_env_vars.get("TORCH_LOGS").map(String::as_str),
        Some("+recompiles")
    );

    // One record from each top-level array is present.
    assert_eq!(artifact.graph_breaks.len(), 1);
    assert_eq!(artifact.recompilations.len(), 1);
    assert_eq!(artifact.compiled_graphs[0].kernel_ids, vec!["k_001"]);
    assert_eq!(artifact.compile_phases.len(), 2);
    assert_eq!(artifact.iterations[0].cache_hit, Some(true));
    assert_eq!(
        artifact.lint_findings[0].pattern_category,
        "in_place_op_on_alias"
    );
    assert_eq!(
        artifact.roofline_predictions[0].bound_type.as_deref(),
        Some("memory")
    );

    // Tool 6 fusion opportunity (ADR-037): pattern + shape + suggested fusion survive deserialize.
    let fusion = &artifact.fusion_opportunities[0];
    assert_eq!(fusion.pattern_id, "A");
    assert_eq!(fusion.confidence.as_deref(), Some("high"));
    assert_eq!(fusion.shape.as_ref().expect("shape present").m, Some(8192));
    assert_eq!(
        fusion.suggested_fusion.as_deref(),
        Some("fold per-row 1/rms into the second GEMM epilogue")
    );

    // Tool 3 divergence finding, with its nested causal attribution (ADR-034).
    let divergence = &artifact.divergences[0];
    assert_eq!(
        divergence.first_divergent_layer.as_deref(),
        Some("decoder.layers.3.mlp.fc2")
    );
    assert_eq!(divergence.max_abs_diff, Some(0.297));
    assert_eq!(divergence.num_layers_compared, 48);
    let attribution = divergence
        .attribution
        .as_ref()
        .expect("attribution present");
    assert!(attribution.attributed);
    assert_eq!(
        attribution.responsible_passes,
        vec!["post_grad_custom_post_pass"]
    );
    assert_eq!(attribution.num_probes, 8);

    // number-typed feature arrives as f64 even though it's written as an integer literal.
    let flops = artifact.kernels[0]
        .features
        .as_ref()
        .unwrap()
        .flops
        .unwrap();
    assert_eq!(flops, 1.2e10);
}

// ── test_kernel_features_num_regs_present ───────────────────────────────────────────────
#[test]
fn test_kernel_features_num_regs_present() {
    let artifact = parse(FULL);
    let features = artifact.kernels[0]
        .features
        .as_ref()
        .expect("kernel features present");
    // Register pressure proxy must survive deserialization (Tool 5's 4th correction).
    assert_eq!(features.num_regs, Some(168));
    assert_eq!(features.n_spills, Some(0));
}

// ── test_fx_nodes_parse_with_ordered_inputs ─────────────────────────────────────────────
#[test]
fn test_fx_nodes_parse_with_ordered_inputs() {
    let artifact = parse(FULL);
    let nodes = &artifact.compiled_graphs[0].nodes;
    assert_eq!(nodes.len(), 4);

    // A required-only node: no inputs, no attrs.
    assert_eq!(nodes[0].id, "n0");
    assert_eq!(nodes[0].op_type, "placeholder");
    assert!(nodes[0].inputs.is_empty());
    assert!(nodes[0].attrs.is_empty());

    // The sub node carries inputs in load-bearing order (n0 before n1) plus an attr.
    let sub = &nodes[2];
    assert_eq!(sub.op_type, "aten.sub");
    assert_eq!(sub.inputs, vec!["n0", "n1"]);
    assert_eq!(sub.attrs.get("alpha").and_then(|v| v.as_i64()), Some(1));

    // Operand order is preserved as a sequence, not normalized to a set: reversing it is a
    // genuinely different node, which is exactly what the WL diff relies on.
    assert_ne!(sub.inputs, vec!["n1", "n0"]);
}

// ── test_fx_nodes_absent_when_not_captured ──────────────────────────────────────────────
#[test]
fn test_fx_nodes_absent_when_not_captured() {
    // A graph written before node capture (the minimal artifact has none) parses fine: the
    // field is optional and defaults to empty rather than erroring.
    let artifact = parse(MINIMAL);
    assert!(artifact.compiled_graphs.is_empty());

    // And an empty `nodes` is omitted on re-serialize (skip_serializing_if), so a Tool 1
    // artifact does not grow a spurious `"nodes":[]`.
    let graph = cls_schema::CompiledGraph {
        graph_id: "g".into(),
        compiled_function_id: None,
        fx_graph_path: None,
        inductor_ir_path: None,
        guard_list: Vec::new(),
        kernel_ids: Vec::new(),
        compile_phases_summary: Default::default(),
        nodes: Vec::new(),
    };
    let json = serde_json::to_string(&graph).unwrap();
    assert!(!json.contains("nodes"));
}

// ── test_divergence_round_trips_and_skips_when_empty ─────────────────────────────────────
#[test]
fn test_divergence_round_trips_and_skips_when_empty() {
    use cls_schema::{Divergence, DivergenceAttribution};

    // A no-divergence finding (first_divergent_layer / max_abs_diff / attribution all absent)
    // round-trips without acquiring or losing fields, and the absent optionals are not emitted.
    let mut artifact = parse(MINIMAL);
    artifact.divergences.push(Divergence {
        divergence_id: "dv_x".into(),
        first_divergent_layer: None,
        max_abs_diff: None,
        num_layers_compared: 12,
        rtol: 1e-3,
        atol: 1e-5,
        suggested_cause: None,
        attribution: None,
    });
    let json = serde_json::to_string(&artifact).expect("serialize");
    let reparsed: ClsArtifact = serde_json::from_str(&json).expect("re-deserialize");
    assert_eq!(artifact, reparsed);
    assert!(!json.contains("first_divergent_layer"));
    assert!(!json.contains("attribution"));

    // An attributed finding carries its nested causal attribution through the cycle (ADR-034).
    artifact.divergences[0].attribution = Some(DivergenceAttribution {
        attributed: true,
        responsible_passes: vec!["epilogue_fusion".into()],
        summary: "divergence removed when inductor pass(es) disabled: epilogue_fusion".into(),
        num_probes: 6,
    });
    let json2 = serde_json::to_string(&artifact).expect("serialize");
    let reparsed2: ClsArtifact = serde_json::from_str(&json2).expect("re-deserialize");
    assert_eq!(artifact, reparsed2);

    // An empty divergences array is omitted entirely (skip_serializing_if), so a non-Tool-3
    // artifact does not grow a spurious "divergences":[].
    let empty_json = serde_json::to_string(&parse(MINIMAL)).expect("serialize");
    assert!(!empty_json.contains("divergences"));
}

// ── test_fusion_opportunity_round_trips_and_skips_when_empty ─────────────────────────────
#[test]
fn test_fusion_opportunity_round_trips_and_skips_when_empty() {
    use cls_schema::{FusionLocation, FusionOpportunity, FusionShape};

    // An artifact without Tool 6 results omits the array entirely (skip_serializing_if), so a
    // pre-Tool-6 artifact does not grow a spurious "fusion_opportunities":[] — and an old artifact
    // read back simply defaults the field to empty (forward compatibility).
    let empty_json = serde_json::to_string(&parse(MINIMAL)).expect("serialize");
    assert!(!empty_json.contains("fusion_opportunities"));

    // A full opportunity round-trips structurally; an absent optional (src_lineno_range) is not
    // emitted.
    let mut artifact = parse(MINIMAL);
    artifact.fusion_opportunities.push(FusionOpportunity {
        pattern_id: "A".into(),
        location: Some(FusionLocation {
            fx_node_ids: vec!["n_1".into(), "n_2".into()],
            src_lineno_range: None,
        }),
        shape: Some(FusionShape {
            m: Some(8192),
            n: Some(11008),
            k0: Some(4096),
            k1: Some(4096),
            dtype: Some("bf16".into()),
        }),
        baseline_hbm_bytes: Some(432_013_312.0),
        fused_hbm_bytes: Some(300_941_312.0),
        estimated_speedup: Some(1.43),
        suggested_fusion: Some("fold per-row 1/rms into the second GEMM epilogue".into()),
        confidence: Some("high".into()),
    });
    let json = serde_json::to_string(&artifact).expect("serialize");
    let reparsed: ClsArtifact = serde_json::from_str(&json).expect("re-deserialize");
    assert_eq!(artifact, reparsed);
    assert!(!json.contains("src_lineno_range"));
}

// ── test_session_round_trip_serde ───────────────────────────────────────────────────────
#[test]
fn test_session_round_trip_serde() {
    let original = parse(FULL);
    let json = serde_json::to_string(&original).expect("serialize");
    let reparsed: ClsArtifact = serde_json::from_str(&json).expect("re-deserialize");
    // Structural equality across a full serialize -> deserialize cycle.
    assert_eq!(original, reparsed);

    // Minimal must also round-trip without acquiring extra fields.
    let minimal = parse(MINIMAL);
    let minimal_json = serde_json::to_string(&minimal).expect("serialize minimal");
    assert_eq!(
        minimal,
        serde_json::from_str(&minimal_json).expect("re-deserialize minimal")
    );
}

// ── test_redaction_policy_enum_serialize ────────────────────────────────────────────────
#[test]
fn test_redaction_policy_enum_serialize() {
    let cases = [
        (RedactionPolicy::DefaultStrict, "\"default-strict\""),
        (RedactionPolicy::Internal, "\"internal\""),
        (RedactionPolicy::Confidential, "\"confidential\""),
        (RedactionPolicy::PublicSafe, "\"public-safe\""),
    ];
    for (policy, wire) in cases {
        assert_eq!(serde_json::to_string(&policy).unwrap(), wire);
        assert_eq!(
            serde_json::from_str::<RedactionPolicy>(wire).unwrap(),
            policy
        );
    }
    // Fail closed: an unknown policy is a hard error, never a silent permissive default.
    assert!(serde_json::from_str::<RedactionPolicy>("\"wide-open\"").is_err());
}

// ── test_unknown_field_handled ──────────────────────────────────────────────────────────
#[test]
fn test_unknown_field_handled() {
    // Inject unknown keys at the top level, the session level, and inside a nested record.
    let mut value: serde_json::Value = serde_json::from_str(FULL).unwrap();
    value["future_top_level_array"] = serde_json::json!([{ "x": 1 }]);
    value["session"]["gpu_arch"] = serde_json::json!("hopper");
    value["graph_breaks"][0]["future_break_field"] = serde_json::json!("anything");
    let json = serde_json::to_string(&value).unwrap();

    let artifact: ClsArtifact = serde_json::from_str(&json).expect("unknown keys must not error");

    // Growth points capture unknown keys so they survive a round trip (D6).
    assert!(artifact.extra.contains_key("future_top_level_array"));
    assert_eq!(
        artifact
            .session
            .extra
            .get("gpu_arch")
            .and_then(|v| v.as_str()),
        Some("hopper")
    );

    // Deeper records accept the unknown key (no error) but do not preserve it.
    let reserialized = serde_json::to_string(&artifact).unwrap();
    assert!(reserialized.contains("future_top_level_array"));
    assert!(reserialized.contains("gpu_arch"));
    assert!(!reserialized.contains("future_break_field"));
}

// ── test_nested_object_key_order_preserved (round-trip determinism, D10) ──────────────────
#[test]
fn test_nested_object_key_order_preserved() {
    // A multi-key object captured at a growth point must re-serialize with its keys in the SAME
    // order (byte-identical round-trip, ADR-021 / D10). Without serde_json's `preserve_order`, a
    // nested Value::Object uses a BTreeMap and alphabetizes — silently breaking the contract and
    // drifting from the Python side, whose dict preserves insertion order. The keys here are
    // deliberately NOT in alphabetical order.
    let json = r#"{
        "schema_version": "0.5.0",
        "session": { "id": "x", "timestamp": "t", "torch_version": "2.6.0", "redaction_policy": "default-strict" },
        "future_options": { "zzz": 1, "aaa": 2, "mmm": 3 }
    }"#;
    let artifact: ClsArtifact = serde_json::from_str(json).expect("unknown keys must not error");
    let reserialized = serde_json::to_string(&artifact).expect("serializes");

    let z = reserialized.find("zzz").expect("zzz present");
    let a = reserialized.find("aaa").expect("aaa present");
    let m = reserialized.find("mmm").expect("mmm present");
    assert!(
        z < a && a < m,
        "nested object keys must keep insertion order (zzz, aaa, mmm), got:\n{reserialized}"
    );
}

// ── extra: constructor + missing-required-field behavior ────────────────────────────────
#[test]
fn test_new_stamps_current_version() {
    let session = parse(MINIMAL).session;
    let artifact = ClsArtifact::new(session);
    assert_eq!(artifact.schema_version, SCHEMA_VERSION);
    assert!(artifact.kernels.is_empty());
}

#[test]
fn test_missing_required_field_fails() {
    // Session without the required `redaction_policy` must fail to deserialize.
    let bad = r#"{ "schema_version": "0.5.0",
                   "session": { "id": "x", "timestamp": "t", "torch_version": "2.6.0" } }"#;
    assert!(serde_json::from_str::<ClsArtifact>(bad).is_err());

    // A bare Session missing `torch_version` likewise fails.
    let bad_session = r#"{ "id": "x", "timestamp": "t", "redaction_policy": "internal" }"#;
    assert!(serde_json::from_str::<Session>(bad_session).is_err());
}
